service/bfg/bfg.go

```diff
@@ -317,6 +319,102 @@ func (s *Server) handleBitcoinBalance(ctx context.Context, bbr *bfgapi.BitcoinBa
 	}, nil
 }
 
+func (s *Server) bitcoinBroadcastWorker(ctxI context.Context, highPriority bool) {
+	log.Tracef("bitcoinBroadcastWorker")
+	defer log.Tracef("bitcoinBroadcastWorker exit")
+
+	defer s.wg.Done()
+
+	for {
+		select {
+		case <-ctxI.Done():
+			return
+		default:
+		}
+
+		ctx, cancel := context.WithTimeout(ctxI, 5*time.Second)
+
+		serializedTx, err := s.db.BtcTransactionBroadcastRequestGetNext(ctx, highPriority)
+		if err != nil {
+			log.Errorf("error getting next broadcast request: %v", err)
+			cancel()
+			continue
+		}
+
+		// if there are no new serialized txs, backoff a bit
+		if serializedTx == nil {
+			cancel()
+			select {
+			case <-time.After(5 * time.Second):
+				continue
+			}
+		}
+
+		rr := bytes.NewReader(serializedTx)
+		mb := wire.MsgTx{}
+		if err := mb.Deserialize(rr); err != nil {
+			log.Errorf("failed to deserialize tx: %v", err)
+			cancel()
+			continue
+		}
+
+		var tl2 *pop.TransactionL2
+		for _, v := range mb.TxOut {
+			tl2, err = pop.ParseTransactionL2FromOpReturn(v.PkScript)
+			if err == nil {
+				break // Found the pop transaction.
+			}
+		}
+
+		if tl2 == nil {
+			log.Errorf("could not find pop tx")
+			cancel()
+			continue
+		}
+
+		publicKeyUncompressed, err := pop.ParsePublicKeyFromSignatureScript(mb.TxIn[0].SignatureScript)
+		if err != nil {
+			log.Errorf("could not parse public key from signature script: %v", err)
+		}
+
+		hash := mb.TxHash()
+
+		if err := s.db.PopBasisInsertPopMFields(ctx, &bfgd.PopBasis{
+			BtcTxId:             hash[:],
+			BtcRawTx:            database.ByteArray(serializedTx),
+			PopMinerPublicKey:   publicKeyUncompressed,
+			L2KeystoneAbrevHash: tl2.L2Keystone.Hash(),
+		}); err != nil {
+			log.Infof("inserting pop basis: %s", err)
+		}
+
+		_, err = s.btcClient.Broadcast(ctx, serializedTx)
+		if err != nil {
+			log.Errorf("broadcast tx: %s", err)
+			err = s.db.BtcTransactionBroadcastRequestSetLastError(ctx, mb.TxID(), err.Error())
+			if err != nil {
+				log.Errorf("could not delete %v", err)
+			}
+			cancel()
+			continue
+		}
+
+		s.metrics.popBroadcasts.Inc()
+
+		log.Infof("hash is %s", hex.EncodeToString(hash[:]))
+
+		err = s.db.BtcTransactionBroadcastRequestConfirmBroadcast(ctx, mb.TxID())
+		if err != nil {
+			log.Errorf("could not confirm broadcast: %v", err)
+			cancel()
+			continue
+		}
+
+		log.Infof("successfully broadcast tx %s, for l2 keystone %s", mb.TxID(), hex.EncodeToString(tl2.L2Keystone.Hash()))
+		cancel()
```

Comment:
I think it would be a good idea to extract this into a separate function, allowing us to `defer cancel()` and use `return`. This will prevent accidental context leaks.

Reply:
ok let me see what I can do 🕵️‍♂️